app/src/org/commcare/activities/connect/ConnectIdBiometricConfigActivity.java

```diff
@@ -116,34 +128,92 @@ public void updateState(BiometricsHelper.ConfigurationStatus fingerprintStatus,
-            finish(true, true);
+
+        if (BiometricsHelper.isFingerprintConfigured(this, biometricManager)) {
+            //Automatically try fingerprint first
+            performFingerprintUnlock();
+        } else if (BiometricsHelper.isPinConfigured(this, biometricManager)) {
+            performPinUnlock();
+        } else if (allowPassword) {
+            performPasswordUnlock();
+        } else {
+            Logger.exception("No unlock method available when trying to unlock ConnectID", new Exception("No unlock option"));
         }
     }
 
-    public void handlePinButton() {
-        BiometricsHelper.ConfigurationStatus pin = BiometricsHelper.checkPinStatus(this, biometricManager);
-        if (pin == BiometricsHelper.ConfigurationStatus.Configured) {
-            finish(true, false);
-        } else if (!BiometricsHelper.configurePin(this)) {
-            finish(true, true);
-        }
+    public void performFingerprintUnlock() {
+        attemptingFingerprint = true;
+        boolean allowOtherOptions = BiometricsHelper.isPinConfigured(this, biometricManager) ||
+                allowPassword;
+        BiometricsHelper.authenticateFingerprint(this, biometricManager, allowOtherOptions, biometricPromptCallbacks);
     }
 
-    public void finish(boolean success, boolean failedEnrollment) {
-        Intent intent = new Intent(getIntent());
+    public void performPasswordUnlock() {
+        finish(false, true, false);
+    }
 
-        BiometricsHelper.ConfigurationStatus fingerprint = BiometricsHelper.checkFingerprintStatus(this,
-                biometricManager);
-        BiometricsHelper.ConfigurationStatus pin = BiometricsHelper.checkPinStatus(this, biometricManager);
-        boolean configured = fingerprint == BiometricsHelper.ConfigurationStatus.Configured ||
-                pin == BiometricsHelper.ConfigurationStatus.Configured;
+    public void performPinUnlock() {
+        BiometricsHelper.authenticatePin(this, biometricManager, biometricPromptCallbacks);
+    }
+
+    private BiometricPrompt.AuthenticationCallback preparePromptCallbacks() {
+        final Context context = this;
+        return new BiometricPrompt.AuthenticationCallback() {
+            @Override
+            public void onAuthenticationError(int errorCode,
+                                              @NonNull CharSequence errString) {
+                super.onAuthenticationError(errorCode, errString);
+                if (attemptingFingerprint) {
+                    attemptingFingerprint = false;
+                    if (BiometricsHelper.isPinConfigured(context, biometricManager) &&
+                            allowPassword) {
+                        //Automatically try password, it's the only option
+                        performPinUnlock();
+                    } else {
+                        //Automatically try password, it's the only option
+                        performPasswordUnlock();
```

Comment:
No password fallback anymore, at this point the user needs to recover their account